leader/leader.go

```diff
@@ -88,241 +87,282 @@ type CallbackFn func(key, prevValue, newValue string)
+	logger := log.WithField("key", key)
+	logger.WithField("peers", l.Client.Endpoints()).Info("Setting up watch.")
 
-		backoff := NewUnlimitedExponentialBackOff()
-		ticker := ebackoff.NewTicker(backoff)
-		var steps int
+	go l.watchLoop(ctx, key, valuesC, logger)
+}
 
-		watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
-		if err != nil {
-			return
-		}
+// AddVoter starts a goroutine that attempts to set the specified key to
+// to the given value with the time-to-live value specified with term.
+// The time-to-live value cannot be less than a second.
+// After successfully setting the key, it attempts to renew the lease for the specified
+// term indefinitely.
+// The method is idempotent and does nothing if invoked multiple times
+func (l *Client) AddVoter(ctx context.Context, key, value string, term time.Duration) {
+	l.once.Do(func() {
+		l.startVoterLoop(key, value, term, true)
+	})
+	select {
+	case l.voterC <- true:
+	case <-ctx.Done():
+	case <-l.closeC:
+	}
+}
 
-		// make sure we always send the first actual value
-		if resp != nil && resp.Node != nil {
-			select {
-			case valuesC <- resp.Node.Value:
-			case <-l.closeC:
-				return
-			}
-		}
+// RemoveVoter stops the voting loop.
+func (l *Client) RemoveVoter(ctx context.Context, key, value string, term time.Duration) {
+	l.once.Do(func() {
+		l.startVoterLoop(key, value, term, false)
+	})
+	select {
+	case l.voterC <- false:
+	case <-ctx.Done():
+	case <-l.closeC:
+	}
+}
 
-		var sentAnything bool
-		for {
+// StepDown makes this participant to pause his attempts to re-elect itself thus giving up its leadership
+func (l *Client) StepDown(ctx context.Context) {
+	select {
+	case l.pauseC <- true:
+	case <-ctx.Done():
+	}
+}
 
-			if watcher == nil {
-				watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
-			}
+// Close stops current operations and releases resources
+func (l *Client) Close() error {
+	// already closed
+	if !atomic.CompareAndSwapUint32(&l.closed, 0, 1) {
+		return nil
+	}
+	close(l.closeC)
+	l.wg.Wait()
+	return nil
+}
 
-			if watcher != nil {
-				resp, err = watcher.Next(ctx)
-				if err == nil {
-					if resp.Node.Value == "" {
-						continue
-					}
-					backoff.Reset()
-				}
-			}
+func (l *Client) startVoterLoop(key, value string, term time.Duration, enabled bool) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-l.closeC
+		cancel()
+	}()
+	go l.voterLoop(ctx, key, value, term, enabled)
+}
 
+func (l *Client) watchLoop(ctx context.Context, key string, valuesC chan string, logger logrus.FieldLogger) {
+	defer l.wg.Done()
+	boff := newBackoff()
+	// maxFailedSteps sets the limit on the number of failed attempts before the watch
+	// is forcibly reset
+	const maxFailedSteps = 10
+	var (
+		api     = client.NewKeysAPI(l.Client)
+		watcher client.Watcher
+	)
+	var err error
+	for {
+		select {
+		case <-time.After(boff.NextBackOff()):
+		case <-l.closeC:
```

Comment:
Should there also be a `case <-ctx.Done()`? Also, can we log when it exits?

Reply:
No since the context is driven by `closeC` which is already queried.
Sure, I'll log the event.